### PR TITLE
Implement basic authentication

### PR DESCRIPTION
**Pull request by krabique**

### CODE REVIEW

File: spec/feature/log_in_spec.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+require 'rails_helper'
+
+RSpec.feature 'SignUpsAndLogIns', type: :feature do
+  let(:name) { Faker::DrWho.character }
+  let(:email) { Faker::Internet.email }
+  let(:password) { Faker::Internet.password }
+
+  scenario 'a visitor can log in' do
+    user = User.create(name: name, email: email, password: password)
```
Comment (roman-dubrovsky) [inline spec/feature/log_in_spec.rb]: let's generate user via [fabrication](https://github.com/paulelliott/fabrication) or [Factory Bot](https://github.com/thoughtbot/factory_bot_rails)

File: app/views/shared/_flash_messages.html.slim [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#flash_messages
```
Comment (roman-dubrovsky) [inline app/views/shared/_flash_messages.html.slim]: let's set up a [rails view](https://github.com/trailblazer/cells) into the project and move flash notifications and navbar into a separate component. Read more https://datarockets.com/blog/object-oriented-views-in-rails

You can do it as a separate task (create task for this in Trello and do it later). But in this case, don't comment about this below.

File: app/views/users/confirmations/new.html.slim [hunk 3]
```diff
@@ -0,0 +1,10 @@
+h2
```
Comment (roman-dubrovsky) [inline app/views/users/confirmations/new.html.slim]: do you need this views? What about removing this?

File: .rubocop.yml [hunk 4]
```diff
@@ -14,5 +14,41 @@ Style/MixinUsage:
   Exclude:
     - bin/setup
     - bin/update
+
 Style/EmptyMethod:
   EnforcedStyle: expanded
+
+Style/ClassAndModuleChildren:
+  EnforcedStyle: compact
+  Exclude:
```
Comment (roman-dubrovsky) [inline .rubocop.yml]: and don't forget to remove `Exclude` block

Comment (roman-dubrovsky) [reply .rubocop.yml]: I'm not sure about `config/application.rb`, but can you remove `app/controllers/users/**/*` from this block?

Comment (krabique) [reply .rubocop.yml]: the reason for excluding `config/application.rb` is that rubocop doesn't like
```
module FinancialPlanner
  class Application < Rails::Application
```
so I try to change it to
```
class FinancialPlanner::Application < Rails::Application
```
but then tests fail with
```
└─ $ ▶ bundle exec rspec

An error occurred while loading ./spec/controllers/home_controller_spec.rb.
Failure/Error: require File.expand_path('../../config/environment', __FILE__)

NameError:
  uninitialized constant FinancialPlanner
# ./config/application.rb:11:in `<top (required)>'
# ./config/environment.rb:4:in `require_relative'
# ./config/environment.rb:4:in `<top (required)>'
# ./spec/rails_helper.rb:6:in `require'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./spec/controllers/home_controller_spec.rb:3:in `require'
# ./spec/controllers/home_controller_spec.rb:3:in `<top (required)>'

An error occurred while loading ./spec/feature/log_in_spec.rb.
Failure/Error: require File.expand_path('../../config/environment', __FILE__)

NameError:
  uninitialized constant FinancialPlanner
# ./config/application.rb:11:in `<top (required)>'
# ./config/environment.rb:4:in `require_relative'
# ./config/environment.rb:4:in `<top (required)>'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./spec/feature/log_in_spec.rb:3:in `<top (required)>'

An error occurred while loading ./spec/feature/sign_up_spec.rb.
Failure/Error: require File.expand_path('../../config/environment', __FILE__)

NameError:
  uninitialized constant FinancialPlanner
# ./config/application.rb:11:in `<top (required)>'
# ./config/environment.rb:4:in `require_relative'
# ./config/environment.rb:4:in `<top (required)>'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./spec/feature/sign_up_spec.rb:3:in `<top (required)>'

An error occurred while loading ./spec/helpers/application_helper_spec.rb.
Failure/Error: require File.expand_path('../../config/environment', __FILE__)

NameError:
  uninitialized constant FinancialPlanner
# ./config/application.rb:11:in `<top (required)>'
# ./config/environment.rb:4:in `require_relative'
# ./config/environment.rb:4:in `<top (required)>'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./spec/helpers/application_helper_spec.rb:3:in `<top (required)>'

An error occurred while loading ./spec/models/user_spec.rb.
Failure/Error: require File.expand_path('../../config/environment', __FILE__)

NameError:
  uninitialized constant FinancialPlanner
# ./config/application.rb:11:in `<top (required)>'
# ./config/environment.rb:4:in `require_relative'
# ./config/environment.rb:4:in `<top (required)>'
# ./spec/rails_helper.rb:6:in `<top (required)>'
# ./spec/models/user_spec.rb:3:in `<top (required)>'
No examples found.


Finished in 0.00029 seconds (files took 3.2 seconds to load)
0 examples, 0 failures, 5 errors occurred outside of examples
```

Comment (roman-dubrovsky) [reply .rubocop.yml]: ok, leave `config/application.rb` in the excluding block

File: app/controllers/users/confirmations_controller.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::ConfirmationsController < Devise::ConfirmationsController
```
[the remark this answers is not in the document]
Comment (krabique) [reply app/controllers/users/confirmations_controller.rb]: Yes, those `users` controllers and views are required to override the default devise's views. I use `simple_form` here to make the input fields and buttons look bootstrapish, for example.

If i remove, say, `sessions_controller.rb`, the log-in page would look like this:
screenshot_20180226_162525

I override it to look like this:
screenshot_20180226_162415

[the remark this answers is not in the document]
Comment (roman-dubrovsky) [reply app/controllers/users/confirmations_controller.rb]: @krabique read more how to do it here https://github.com/plataformatec/devise#configuring-views

I suggest you run `rails generate devise:views` command which generates directory `app/views/devise/` for you where you can change some of the views. 

You can run this command and then move all necessary files from `app/views/users` (but don't forget to remove `app/views/users` after this)

File: app/controllers/users/registrations_controller.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
+# frozen_string_literal: true
+
+class Users::RegistrationsController < Devise::RegistrationsController
+  before_action :configure_sign_up_params, only: [:create]
+  before_action :configure_account_update_params, only: [:update]
+
+  # GET /resource/sign_up
```
Comment (roman-dubrovsky) [inline app/controllers/users/registrations_controller.rb]: can you remove all commented code?

File: config/environment.rb [hunk 7]
```diff
@@ -5,3 +5,13 @@
 
 # Initialize the Rails application.
 Rails.application.initialize!
+
+ActionMailer::Base.smtp_settings = {
+  user_name: ENV['MAIL_USERNAME'],
```
Comment (roman-dubrovsky) [inline config/environment.rb]: you save credentials for mail in the `.env` file, don't you? But the person who forks your project will forget to set these variables.  So can you add `.env.sample` file with empty variables and add the line to `bin/setup` which creates  `.env` from sample file?

**Pull request merged**